File: packages/constants/src/order.ts
/**
 * 订单相关常量
 */

/** 订单状态 */
export const ORDER_STATUS = {
  /** 待付款 */
  PENDING_PAYMENT: 'pending_payment',
  /** 待确认 */
  PENDING_CONFIRM: 'pending_confirm',
  /** 已确认 */
  CONFIRMED: 'confirmed',
  /** 配送中 */
  DELIVERING: 'delivering',
  /** 已完成 */
  COMPLETED: 'completed',
  /** 已取消 */
  CANCELLED: 'cancelled',
} as const;

/** 订单状态文本 */
export const ORDER_STATUS_TEXT: Record<string, string> = {
  [ORDER_STATUS.PENDING_PAYMENT]: '待付款',
  [ORDER_STATUS.PENDING_CONFIRM]: '待确认',
  [ORDER_STATUS.CONFIRMED]: '已确认',
  [ORDER_STATUS.DELIVERING]: '配送中',
  [ORDER_STATUS.COMPLETED]: '已完成',
  [ORDER_STATUS.CANCELLED]: '已取消',
};

/** 订单状态颜色 */
export const ORDER_STATUS_COLOR: Record<string, string> = {
  [ORDER_STATUS.PENDING_PAYMENT]: '#f59e0b', // 橙色
  [ORDER_STATUS.PENDING_CONFIRM]: '#3b82f6', // 蓝色
  [ORDER_STATUS.CONFIRMED]: '#8b5cf6', // 紫色
  [ORDER_STATUS.DELIVERING]: '#06b6d4', // 青色
  [ORDER_STATUS.COMPLETED]: '#22c55e', // 绿色
  [ORDER_STATUS.CANCELLED]: '#6b7280', // 灰色
};

/** 支付状态 */
export const PAYMENT_STATUS = {
  /** 未支付 */
  UNPAID: 'unpaid',
  /** 已支付 */
  PAID: 'paid',
  /** 已退款 */
  REFUNDED: 'refunded',
} as const;

/** 支付状态文本 */
export const PAYMENT_STATUS_TEXT: Record<string, string> = {
  [PAYMENT_STATUS.UNPAID]: '未支付',
  [PAYMENT_STATUS.PAID]: '已支付',
  [PAYMENT_STATUS.REFUNDED]: '已退款',
};

/** 支付方式 */
export const PAYMENT_METHOD = {
  /** 微信支付 */
  WECHAT: 'wechat',
  /** 支付宝 */
  ALIPAY: 'alipay',
} as const;

/** 支付方式文本 */
export const PAYMENT_METHOD_TEXT: Record<string, string> = {
  [PAYMENT_METHOD.WECHAT]: '微信支付',
  [PAYMENT_METHOD.ALIPAY]: '支付宝',
};

/** 订单来源 */
export const ORDER_SOURCE = {
  /** APP */
  APP: 'app',
  /** Web */
  WEB: 'web',
  /** H5 */
  H5: 'h5',
} as const;

/** 订单来源文本 */
export const ORDER_SOURCE_TEXT: Record<string, string> = {
  [ORDER_SOURCE.APP]: 'APP',
  [ORDER_SOURCE.WEB]: 'Web端',
  [ORDER_SOURCE.H5]: 'H5',
};

/** 取消人类型 */
export const CANCELLED_BY = {
  /** 门店 */
  STORE: 'store',
  /** 供应商 */
  SUPPLIER: 'supplier',
  /** 管理员 */
  ADMIN: 'admin',
  /** 系统 */
  SYSTEM: 'system',
} as const;

/** 取消人类型文本 */
export const CANCELLED_BY_TEXT: Record<string, string> = {
  [CANCELLED_BY.STORE]: '门店',
  [CANCELLED_BY.SUPPLIER]: '供应商',
  [CANCELLED_BY.ADMIN]: '管理员',
  [CANCELLED_BY.SYSTEM]: '系统',
};

/** 取消申请状态 */
export const CANCEL_REQUEST_STATUS = {
  /** 待处理 */
  PENDING: 'pending',
  /** 已通过 */
  APPROVED: 'approved',
  /** 已拒绝 */
  REJECTED: 'rejected',
} as const;

/** 取消申请状态文本 */
export const CANCEL_REQUEST_STATUS_TEXT: Record<string, string> = {
  [CANCEL_REQUEST_STATUS.PENDING]: '待处理',
  [CANCEL_REQUEST_STATUS.APPROVED]: '已通过',
  [CANCEL_REQUEST_STATUS.REJECTED]: '已拒绝',
};

/** 订单状态流转映射（当前状态可以转换到的状态） */
export const ORDER_STATUS_FLOW: Record<string, string[]> = {
  [ORDER_STATUS.PENDING_PAYMENT]: [ORDER_STATUS.PENDING_CONFIRM, ORDER_STATUS.CANCELLED],
  [ORDER_STATUS.PENDING_CONFIRM]: [ORDER_STATUS.CONFIRMED, ORDER_STATUS.CANCELLED],
  [ORDER_STATUS.CONFIRMED]: [ORDER_STATUS.DELIVERING, ORDER_STATUS.CANCELLED],
  [ORDER_STATUS.DELIVERING]: [ORDER_STATUS.COMPLETED],
  [ORDER_STATUS.COMPLETED]: [],
  [ORDER_STATUS.CANCELLED]: [ORDER_STATUS.PENDING_CONFIRM], // 恢复订单
};

File: apps/backend/models/order.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// OrderStatus represents order status types
type OrderStatus string

const (
	OrderStatusPendingPayment OrderStatus = "pending_payment"
	OrderStatusUnpaid         OrderStatus = "pending_payment" // Alias for pending_payment
	OrderStatusPendingConfirm OrderStatus = "pending_confirm"
	OrderStatusConfirmed      OrderStatus = "confirmed"
	OrderStatusDelivering     OrderStatus = "delivering"
	OrderStatusCompleted      OrderStatus = "completed"
	OrderStatusCancelled      OrderStatus = "cancelled"
)

// PaymentStatus represents payment status types
type PaymentStatus string

const (
	PaymentStatusUnpaid        PaymentStatus = "unpaid"
	PaymentStatusPaid          PaymentStatus = "paid"
	PaymentStatusRefunded      PaymentStatus = "refunded"
	PaymentStatusPartialRefund PaymentStatus = "partial_refund"
)

// PaymentMethod represents payment method types
type PaymentMethod string

const (
	PaymentMethodWechat PaymentMethod = "wechat"
	PaymentMethodAlipay PaymentMethod = "alipay"
)

// OrderSource represents order source types
type OrderSource string

const (
	OrderSourceApp OrderSource = "app"
	OrderSourceWeb OrderSource = "web"
	OrderSourceH5  OrderSource = "h5"
)

// CancelledByType represents who cancelled the order
type CancelledByType string

const (
	CancelledByStore    CancelledByType = "store"
	CancelledBySupplier CancelledByType = "supplier"
	CancelledByAdmin    CancelledByType = "admin"
	CancelledBySystem   CancelledByType = "system"
)

// Order represents the orders table
type Order struct {
	ID                   uint64           `gorm:"primaryKey;autoIncrement" json:"id"`
	OrderNo              string           `gorm:"type:varchar(30);uniqueIndex;not null" json:"order_no"`
	StoreID              uint64           `gorm:"index;not null" json:"store_id"`
	SupplierID           uint64           `gorm:"index;not null" json:"supplier_id"`
	GoodsAmount          float64          `gorm:"type:decimal(10,2);not null" json:"goods_amount"`
	ServiceFee           float64          `gorm:"type:decimal(10,2);default:0" json:"service_fee"`
	TotalAmount          float64          `gorm:"type:decimal(10,2);not null" json:"total_amount"`
	SupplierAmount       float64          `gorm:"type:decimal(10,2)" json:"supplier_amount"`
	MarkupTotal          float64          `gorm:"type:decimal(10,2);default:0" json:"markup_total"`
	ItemCount            int              `json:"item_count"`
	Status               OrderStatus      `gorm:"type:enum('pending_payment','pending_confirm','confirmed','delivering','completed','cancelled')" json:"status"`
	PaymentStatus        PaymentStatus    `gorm:"type:enum('unpaid','paid','refunded','partial_refund');default:'unpaid'" json:"payment_status"`
	PaymentMethod        *PaymentMethod   `gorm:"type:enum('wechat','alipay')" json:"payment_method,omitempty"`
	PaymentTime          *time.Time       `json:"payment_time,omitempty"`
	PaymentNo            *string          `gorm:"type:varchar(50)" json:"payment_no,omitempty"`
	OrderSource          OrderSource      `gorm:"type:enum('app','web','h5')" json:"order_source"`
	DeliveryProvince     *string          `gorm:"type:varchar(50)" json:"delivery_province,omitempty"`
	DeliveryCity         *string          `gorm:"type:varchar(50)" json:"delivery_city,omitempty"`
	DeliveryDistrict     *string          `gorm:"type:varchar(50)" json:"delivery_district,omitempty"`
	DeliveryAddress      *string          `gorm:"type:varchar(200)" json:"delivery_address,omitempty"`
	DeliveryContact      *string          `gorm:"type:varchar(50)" json:"delivery_contact,omitempty"`
	DeliveryPhone        *string          `gorm:"type:varchar(20)" json:"delivery_phone,omitempty"`
	ExpectedDeliveryDate *time.Time       `gorm:"type:date" json:"expected_delivery_date,omitempty"`
	ActualDeliveryTime   *time.Time       `json:"actual_delivery_time,omitempty"`
	Remark               *string          `gorm:"type:varchar(500)" json:"remark,omitempty"`
	SupplierRemark       *string          `gorm:"type:varchar(500)" json:"supplier_remark,omitempty"`
	CancelReason         *string          `gorm:"type:varchar(200)" json:"cancel_reason,omitempty"`
	CancelledBy          *CancelledByType `gorm:"type:enum('store','supplier','admin','system')" json:"cancelled_by,omitempty"`
	CancelledByID        *uint64          `json:"cancelled_by_id,omitempty"`
	CancelledAt          *time.Time       `json:"cancelled_at,omitempty"`
	RestoredAt           *time.Time       `json:"restored_at,omitempty"`
	ConfirmedAt          *time.Time       `json:"confirmed_at,omitempty"`
	DeliveringAt         *time.Time       `json:"delivering_at,omitempty"`
	CompletedAt          *time.Time       `json:"completed_at,omitempty"`
	CreatedAt            time.Time        `json:"created_at"`
	UpdatedAt            time.Time        `json:"updated_at"`
	DeletedAt            gorm.DeletedAt   `gorm:"index" json:"-"`

	// Relationships
	Store      *Store       `gorm:"foreignKey:StoreID" json:"store,omitempty"`
	Supplier   *Supplier    `gorm:"foreignKey:SupplierID" json:"supplier,omitempty"`
	OrderItems []*OrderItem `gorm:"foreignKey:OrderID" json:"order_items,omitempty"`
}

// TableName specifies the table name for Order
func (Order) TableName() string {
	return "orders"
}

// BeforeCreate hook to generate order number
func (o *Order) BeforeCreate(tx *gorm.DB) error {
	if o.OrderNo == "" {
		o.OrderNo = generateOrderNo()
	}
	if o.Status == "" {
		o.Status = OrderStatusPendingPayment
	}
	if o.PaymentStatus == "" {
		o.PaymentStatus = PaymentStatusUnpaid
	}
	return nil
}

// CanCancel checks if the order can be cancelled
func (o *Order) CanCancel(userType string) bool {
	// System and admin can cancel any order
	if userType == "admin" || userType == "system" {
		return o.Status != OrderStatusCancelled && o.Status != OrderStatusCompleted
	}

	// Store can cancel within 1 hour if order is pending confirm
	if userType == "store" {
		if o.Status == OrderStatusPendingConfirm {
			timeSinceCreated := time.Since(o.CreatedAt)
			return timeSinceCreated.Hours() <= 1
		}
	}

	// Supplier can cancel before delivering
	if userType == "supplier" {
		return o.Status == OrderStatusPendingConfirm || o.Status == OrderStatusConfirmed
	}

	return false
}

// IsActive checks if the order is active (not cancelled)
func (o *Order) IsActive() bool {
	return o.Status != OrderStatusCancelled
}

// IsPaid checks if the order is paid
func (o *Order) IsPaid() bool {
	return o.PaymentStatus == PaymentStatusPaid
}

// NeedsPayment checks if the order needs payment
func (o *Order) NeedsPayment() bool {
	return o.Status == OrderStatusPendingPayment && o.PaymentStatus == PaymentStatusUnpaid
}
